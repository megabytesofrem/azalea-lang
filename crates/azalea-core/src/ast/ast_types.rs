use std::hash::Hash;

use crate::parse::span::Span;

use super::{Expr, Stmt};

#[derive(Debug, Clone)]
pub enum Ty {
    Int,
    Float,
    String,
    Bool,
    Unit,

    /// A hole in the type system
    Any,

    /// Represents a type that we don't know yet, will be resolved later on
    Unresolved,

    /// A type variable is a placeholder for a type that can be replaced with any type.
    /// For example, in the type `List[A]`, `A` is a type variable.
    Var(String),

    /// A type constructor is a type that is _parameterized_ by other types.
    /// The following are valid type constructors:
    /// - `List[A]`
    /// - `List[Int]`
    ///
    /// This is not a type constructor:
    /// - `String` ‚ùå
    TypeCons(String, Vec<Ty>),

    /// A type constrained by a type class.
    /// An example of a constraint is `Num[A]`, which means that `A` must be a numeric type.
    /// This is used to represent type classes in the type system.
    Constraint(String, Vec<Ty>),

    /// Maps directly to JavaScript arrays
    Array(Box<Ty>),

    /// Allows us to pass functions around as values
    Fn(Box<Function>),

    /// Record and enum types
    Record(Box<Record>),
    Enum(Box<Enum>),

    /// A universally quantified polymorphic type, e.g. `forall a. a -> a`
    ForAll(Vec<String>, Box<Ty>),
}

// Implement PartialEq, Eq and Hash for `Ty` since it's needed in
// typecheck.

impl PartialEq for Ty {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Ty::Var(a), Ty::Var(b)) => a == b,
            (Ty::TypeCons(a, _), Ty::TypeCons(b, _)) => a == b,
            (Ty::Array(a), Ty::Array(b)) => a == b,
            (Ty::Fn(a), Ty::Fn(b)) => a == b,
            (Ty::Record(a), Ty::Record(b)) => a == b,
            (Ty::Enum(a), Ty::Enum(b)) => a == b,

            _ => core::mem::discriminant(self) == core::mem::discriminant(other),
        }
    }
}

impl Eq for Ty {}

impl Hash for Ty {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        core::mem::discriminant(self).hash(state);
    }
}

/// A pair of a name and a type
pub type TypedPair = (String, Ty);

/// Records are semantically equivalent to JavaScript objects
#[derive(Debug, Clone, PartialEq)]
pub struct Record {
    pub name: String,
    pub fields: Vec<(String, Ty)>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Enum {
    pub name: String,
    pub variants: Vec<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Function {
    pub name: String,
    pub args: Vec<(String, Ty)>,

    // Generic type parameters, e.g. `fn foo[T](x: T): T`
    pub type_params: Vec<String>,

    pub return_ty: Ty,

    // Optional body, if this is used in a function declaration
    pub body: Option<Vec<Span<Stmt>>>,
    pub body_expr: Option<Box<Span<Expr>>>, // If this is a lambda expression
}

#[derive(Debug, Clone, PartialEq)]
pub struct RecordExpr {
    pub name: String,
    pub fields: Vec<(String, Expr)>,
}

impl Record {
    pub fn to_type(&self) -> Ty {
        Ty::Record(Box::new(self.clone()))
    }
}

impl Enum {
    pub fn to_type(&self) -> Ty {
        Ty::Enum(Box::new(self.clone()))
    }
}

impl Function {
    pub fn to_type(&self) -> Ty {
        Ty::Fn(Box::new(self.clone()))
    }

    pub fn new_with_empty(name: String, args: Vec<(String, Ty)>, return_ty: Ty) -> Self {
        Self {
            name,
            args,
            type_params: vec![],
            return_ty,
            body: None,
            body_expr: None,
        }
    }

    pub fn new_with_stmts(
        name: String,
        args: Vec<(String, Ty)>,
        return_ty: Ty,
        body: Vec<Span<Stmt>>,
    ) -> Self {
        Self {
            name,
            args,
            type_params: vec![],
            return_ty,
            body: Some(body),
            body_expr: None,
        }
    }

    pub fn new_with_expr(
        name: String,
        args: Vec<(String, Ty)>,
        return_ty: Ty,
        body_expr: Box<Span<Expr>>,
    ) -> Self {
        Self {
            name,
            args,
            type_params: vec![],
            return_ty,
            body: None,
            body_expr: Some(body_expr),
        }
    }
}
