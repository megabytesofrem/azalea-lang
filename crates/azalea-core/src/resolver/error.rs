use crate::{ast::ast_types::Ty, error::error_report::ErrorReport, lexer::SourceLoc};

#[derive(Debug, Clone)]
pub enum SemanticError {
    UnificationError {
        message: String,
        location: SourceLoc,
    },

    UndefinedVariable(String),
    RedefinedVariable(String),
    UndefinedFunction(String),
    EmptyScope,

    OccursCheck {
        location: SourceLoc,
    },
    TypeMismatch {
        expected: Ty,
        found: Ty,
        location: SourceLoc,
    },

    ExpectedLambda {
        location: SourceLoc,
    },

    UndefinedFieldOrVariant {
        field: String,
        structure: String,
        location: SourceLoc,
    },

    ArityMismatch {
        expected: usize,
        found: usize,
        location: SourceLoc,
    },

    InvalidAssignment {
        location: SourceLoc,
    },
    InvalidTarget {
        location: SourceLoc,
    },
    NoMatch,
}

impl SemanticError {
    pub fn report(&self) -> ErrorReport {
        match self {
            SemanticError::UnificationError { message, location } => ErrorReport::new(
                format!("Unification error: {}", message),
                "Type error".to_string(),
                location.clone(),
            ),

            SemanticError::UndefinedVariable(name) => ErrorReport::new(
                format!("Undefined variable: {}", name),
                "Semantic error".to_string(),
                SourceLoc::default(),
            ),
            SemanticError::RedefinedVariable(name) => ErrorReport::new(
                format!("Redefined variable: {}", name),
                "Semantic error".to_string(),
                SourceLoc::default(),
            ),
            SemanticError::UndefinedFunction(name) => ErrorReport::new(
                format!("Undefined function: {}", name),
                "Semantic error".to_string(),
                SourceLoc::default(),
            ),
            SemanticError::EmptyScope => ErrorReport::new(
                "Attempted to pop an empty scope".to_string(),
                "Semantic error".to_string(),
                SourceLoc::default(),
            ),

            SemanticError::OccursCheck { location } => ErrorReport::new(
                "Occurs check failed".to_string(),
                "Type error".to_string(),
                location.clone(),
            )
            .with_hint("Check for infinite types or circular type references".to_string()),

            SemanticError::TypeMismatch {
                expected,
                found,
                location,
            } => ErrorReport::new(
                format!("Type mismatch: expected {:?}, found {:?}", expected, found),
                "Type error".to_string(),
                location.clone(),
            ),

            SemanticError::ExpectedLambda { location } => ErrorReport::new(
                "Expected a lambda expression".to_string(),
                "Semantic error".to_string(),
                location.clone(),
            )
            .with_hint("Expected a lambda in the format of: \\(params) -> expr".to_string()),

            SemanticError::UndefinedFieldOrVariant {
                field,
                structure,
                location,
            } => ErrorReport::new(
                format!("Undefined field or variant: {} in {}", field, structure),
                "Semantic error".to_string(),
                location.clone(),
            ),

            SemanticError::ArityMismatch {
                expected,
                found,
                location,
            } => ErrorReport::new(
                format!("Arity mismatch: expected {}, found {}", expected, found),
                "Semantic error".to_string(),
                location.clone(),
            ),

            SemanticError::InvalidAssignment { location } => ErrorReport::new(
                "Invalid assignment target".to_string(),
                "Semantic error".to_string(),
                location.clone(),
            )
            .with_hint("Assignment target must be a variable or member access".to_string()),

            SemanticError::InvalidTarget { location } => ErrorReport::new(
                "Invalid target".to_string(),
                "Semantic error".to_string(),
                location.clone(),
            ),

            SemanticError::NoMatch => ErrorReport::new(
                "No matching branch found".to_string(),
                "Semantic error".to_string(),
                SourceLoc::default(),
            ),
        }
    }
}
