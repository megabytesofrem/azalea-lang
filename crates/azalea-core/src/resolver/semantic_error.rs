use crate::{ast::ast_types::Ty, lexer::SourceLoc};

#[derive(Debug, Clone)]
pub enum SemanticError {
    UnificationError(String),

    UndefinedVariable(String),
    RedefinedVariable(String),
    UndefinedFunction(String),
    EmptyScope,

    OccursCheck {
        location: SourceLoc,
    },
    TypeMismatch {
        expected: Ty,
        found: Ty,
        location: SourceLoc,
    },

    ExpectedLambda {
        location: SourceLoc,
    },
}
