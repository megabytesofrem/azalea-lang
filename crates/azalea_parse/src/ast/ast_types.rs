use std::hash::Hash;

use crate::span::Span;

use super::{Block, Stmt, ast::Expr};

#[derive(Debug, Clone)]
pub enum Ty {
    Int,
    Float,
    String,
    Bool,
    Unit,

    /// Represents a type that we don't know yet, will be resolved later on
    UnknownForNow,

    Var(String),

    /// An instantiated type. Instantiated types *may* or *may not* be generic.
    /// The following are all instantiated types:
    /// - `string`
    /// - `List[string]`
    /// - `List[A]`: Do note that `A` is a type variable, but for the purposes of simplicity in
    /// our type system we treat it as an Instantiated type.
    Instantiated(String, Vec<Ty>),

    /// Maps directly to JavaScript arrays
    Array(Box<Ty>),

    /// Allows us to pass functions around as values
    Fn(Box<Function>),

    /// Record and enum types
    Record(Box<Record>),
    Enum(Box<Enum>),
}

// Implement PartialEq, Eq and Hash for `Ty` since it's needed in
// typecheck.

impl PartialEq for Ty {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Ty::Var(a), Ty::Var(b)) => a == b,
            (Ty::Instantiated(a, _), Ty::Instantiated(b, _)) => a == b,
            (Ty::Array(a), Ty::Array(b)) => a == b,
            (Ty::Fn(a), Ty::Fn(b)) => a == b,
            (Ty::Record(a), Ty::Record(b)) => a == b,
            (Ty::Enum(a), Ty::Enum(b)) => a == b,

            _ => core::mem::discriminant(self) == core::mem::discriminant(other),
        }
    }
}

impl Eq for Ty {}

impl Hash for Ty {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        core::mem::discriminant(self).hash(state);
    }
}

/// A pair of a name and a type
pub type TypedPair = (String, Ty);

/// Records are semantically equivalent to JavaScript objects
#[derive(Debug, Clone, PartialEq)]
pub struct Record {
    pub name: String,
    pub fields: Vec<(String, Ty)>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Enum {
    pub name: String,
    pub variants: Vec<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Function {
    pub name: String,
    pub args: Vec<(String, Ty)>,
    pub return_ty: Ty,

    // Optional body, if this is used in a function declaration
    pub body: Option<Vec<Span<Stmt>>>,
    pub body_expr: Option<Box<Span<Expr>>>, // If this is a lambda expression
}

impl Record {
    pub fn to_type(&self) -> Ty {
        Ty::Record(Box::new(self.clone()))
    }
}

impl Enum {
    pub fn to_type(&self) -> Ty {
        Ty::Enum(Box::new(self.clone()))
    }
}

impl Function {
    pub fn to_type(&self) -> Ty {
        Ty::Fn(Box::new(self.clone()))
    }

    pub fn new_with_empty(name: String, args: Vec<(String, Ty)>, return_ty: Ty) -> Self {
        Self {
            name,
            args,
            return_ty,
            body: None,
            body_expr: None,
        }
    }

    pub fn new_with_stmts(
        name: String,
        args: Vec<(String, Ty)>,
        return_ty: Ty,
        body: Vec<Span<Stmt>>,
    ) -> Self {
        Self {
            name,
            args,
            return_ty,
            body: Some(body),
            body_expr: None,
        }
    }

    pub fn new_with_expr(
        name: String,
        args: Vec<(String, Ty)>,
        return_ty: Ty,
        body_expr: Box<Span<Expr>>,
    ) -> Self {
        Self {
            name,
            args,
            return_ty,
            body: None,
            body_expr: Some(body_expr),
        }
    }
}
