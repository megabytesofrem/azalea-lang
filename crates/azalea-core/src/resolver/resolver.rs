use std::{collections::HashMap, vec};

use crate::ast::ast_types::Ty;
use crate::{resolver::Return, resolver::error::SemanticError};

#[derive(Debug, Clone)]
pub struct Scope {
    pub variables: HashMap<String, Ty>,
    pub functions: HashMap<String, Ty>,
}

impl Scope {
    pub fn new() -> Self {
        Self {
            variables: HashMap::new(),
            functions: HashMap::new(),
        }
    }

    pub fn define_variable(&mut self, name: String, ty: Ty) -> Return<()> {
        if self.variables.contains_key(&name) {
            return Err(SemanticError::RedefinedVariable(name));
        }
        self.variables.insert(name, ty);
        Ok(())
    }

    pub fn lookup_variable(&self, name: &str) -> Option<&Ty> {
        self.variables.get(name)
    }

    /// Define a function in this scope
    pub fn define_function(&mut self, name: String, ty: Ty) -> Return<()> {
        if self.functions.contains_key(&name) {
            return Err(SemanticError::RedefinedVariable(name));
        }
        self.functions.insert(name, ty);
        Ok(())
    }

    /// Look up a function in this scope
    pub fn lookup_function(&self, name: &str) -> Option<&Ty> {
        self.functions.get(name)
    }
}

#[derive(Debug, Clone)]
pub struct Resolver {
    scopes: Vec<Scope>,
}

impl Resolver {
    pub fn new() -> Self {
        let global_scope = Scope::new();
        Self {
            scopes: vec![global_scope],
        }
    }

    /// Push a new scope onto the scope stack (for entering blocks, functions, etc.)
    pub fn push_scope(&mut self) {
        self.scopes.push(Scope::new());
    }

    /// Pop the current scope from the scope stack (for exiting blocks, functions, etc.)
    pub fn pop_scope(&mut self) -> Return<()> {
        if self.scopes.len() <= 1 {
            return Err(SemanticError::EmptyScope);
        }
        self.scopes.pop();
        Ok(())
    }

    /// Define a variable in the current scope
    pub fn define_variable(&mut self, name: String, ty: Ty) -> Return<()> {
        if let Some(current_scope) = self.scopes.last_mut() {
            current_scope.define_variable(name, ty)
        } else {
            Err(SemanticError::EmptyScope)
        }
    }

    /// Look up a variable in the scope chain (from innermost to outermost)
    pub fn lookup_variable(&self, name: &str) -> Option<&Ty> {
        // Search from innermost scope to outermost
        for scope in self.scopes.iter().rev() {
            if let Some(ty) = scope.lookup_variable(name) {
                return Some(ty);
            }
        }
        None
    }

    /// Check if a variable is defined in any scope
    pub fn is_variable_defined(&self, name: &str) -> bool {
        self.lookup_variable(name).is_some()
    }

    /// Get the current scope depth (0 = global scope)
    pub fn scope_depth(&self) -> usize {
        self.scopes.len() - 1
    }

    /// Define a function in the current scope
    pub fn define_function(&mut self, name: String, ty: Ty) -> Return<()> {
        if let Some(current_scope) = self.scopes.last_mut() {
            current_scope.define_function(name, ty)
        } else {
            Err(SemanticError::EmptyScope)
        }
    }

    /// Look up a function in the scope chain (from innermost to outermost)
    pub fn lookup_function(&self, name: &str) -> Option<&Ty> {
        // Search from innermost scope to outermost
        for scope in self.scopes.iter().rev() {
            if let Some(ty) = scope.lookup_function(name) {
                return Some(ty);
            }
        }
        None
    }

    /// Check if a function is defined in any scope
    pub fn is_function_defined(&self, name: &str) -> bool {
        self.lookup_function(name).is_some()
    }

    /// Define a variable in the current scope, allowing redefinition (for let statements)
    pub fn define_or_redefine_variable(&mut self, name: String, ty: Ty) {
        if let Some(current_scope) = self.scopes.last_mut() {
            current_scope.variables.insert(name, ty);
        }
    }

    /// Clear all scopes and reset to a single global scope
    pub fn clear(&mut self) {
        self.scopes.clear();
        self.scopes.push(Scope::new());
    }
}
