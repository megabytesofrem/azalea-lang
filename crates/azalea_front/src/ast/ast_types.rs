use super::ast::Expr;

#[derive(Debug, Clone)]
pub enum Ty {
    Int,
    Float,
    String,
    Bool,
    Unit,
    User(String),
    Function(Box<Function>),
    Record(Box<Record>),
}

/// Records are semantically equivalent to JavaScript objects
#[derive(Debug, Clone)]
pub struct Record {
    pub name: String,
    pub fields: Vec<(String, Ty)>,
}

#[derive(Debug, Clone)]
pub struct Enum {
    pub name: String,
    pub variants: Vec<(String, Option<Ty>)>,
}

#[derive(Debug, Clone)]
pub struct Function {
    pub name: String,
    pub args: Vec<(String, Ty)>,
    pub return_ty: Ty,

    pub body: Vec<Expr>,
}
