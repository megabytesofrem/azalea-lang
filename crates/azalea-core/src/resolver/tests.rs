use crate::ast::ast_types::Ty;
use crate::resolver;
use crate::resolver::error::SemanticError;
use crate::resolver::resolver::Resolver;
use insta::assert_snapshot;

macro_rules! assert_define {
    ($r:expr, $name:expr, $ty:expr) => {{
        if !$r.is_variable_defined($name) {
            $r.define_or_redefine_variable($name.to_string(), $ty.clone());
            let result_str = format!("RESULT:\n\n'{}' defined with type: {:#?}", $name, $ty);
            assert_snapshot!(format!("{}", result_str));
        } else {
            let result_str = format!("ERROR:\n\n'{}' is already defined", $name);
            assert_snapshot!(format!("{}", result_str));
        }
    }};
}

macro_rules! assert_error {
    ($result:expr, $expect_err:pat) => {{
        match $result {
            Ok(_) => panic!("Expected an error, but got success"),
            Err(err) => {
                let report = report_error(err.clone());
                assert!(matches!(err, $expect_err));
                assert_snapshot!(format!("ERROR:\n\n{}", report));
            }
        }
    }};
}

macro_rules! assert_lookup {
    ($r:expr, $name:expr) => {{
        let lookup = $r.lookup_variable($name);
        let result_str = lookup.map_or_else(
            || format!("ERROR:\n\nVariable '{}' not found", $name),
            |ty| format!("RESULT:\n\n'{}' has type: {:#?}", $name, ty),
        );
        let snapshot = format!("{}", result_str);
        assert_snapshot!(snapshot);
        lookup
    }};
}

fn define_function(
    resolver: &mut Resolver,
    name: &str,
    args: Vec<(String, Ty)>,
    return_ty: Ty,
) -> Result<(), SemanticError> {
    let function_ty = Ty::Fn(Box::new(crate::ast::ast_types::Function {
        name: name.to_string(),
        args,
        return_ty,
        body: None,
        body_expr: None,
    }));

    resolver.define_function(name.to_string(), function_ty)?;
    Ok(())
}

fn report_error_source(err: SemanticError) -> String {
    let mut report = err.report();
    report.source = "<source>".to_string();
    report.show_pretty_source()
}

fn report_error(err: SemanticError) -> String {
    let mut report = err.report();
    report.source = "<source>".to_string();
    report.show_pretty_message()
}

#[test]
fn basic_variable_definition() {
    let mut resolver = Resolver::new();

    // Define a variable in global scope
    assert_define!(resolver, "x", Ty::Int);
    assert_lookup!(resolver, "x");
}

#[test]
fn errors_on_redefinition() {
    let mut resolver = Resolver::new();

    assert_define!(resolver, "x", Ty::Int);

    // Attempt to redefine the variable
    assert_define!(resolver, "x", Ty::String);
}

#[test]
fn valid_nested_scopes() {
    let mut resolver = Resolver::new();

    // Define variable in global scope
    assert_define!(resolver, "x", Ty::Int);

    // Push new scope
    resolver.push_scope();
    assert_define!(resolver, "y", Ty::String);

    // Pop scope
    resolver.pop_scope().unwrap();

    assert_snapshot!(format!("STATE: {:#?}", resolver));
}

#[test]
fn variable_shadowing() {
    let mut resolver = Resolver::new();

    // Define variable in global scope
    assert_define!(resolver, "x", Ty::Int);

    // Push new scope and shadow the variable
    resolver.push_scope();
    assert_define!(resolver, "x", Ty::String);

    // Pop scope
    resolver.pop_scope().unwrap();

    assert_snapshot!(format!("STATE: {:#?}", resolver));
}

#[test]
fn errors_on_empty_scope() {
    let mut resolver = Resolver::new();

    // Attempt to pop the global scope
    let result = resolver.pop_scope();
    assert_error!(result, SemanticError::EmptyScope);
}

#[test]
fn undefined_variable() {
    let resolver = Resolver::new();

    // Snapshot the lookup result
    assert_error!(
        resolver
            .lookup_variable("undefined_var")
            .ok_or(SemanticError::UndefinedVariable(
                "undefined_var".to_string()
            )),
        SemanticError::UndefinedVariable(_)
    );
}

#[test]
fn function_definition_and_lookup() {
    let mut resolver = Resolver::new();

    // Define a function in global scope
    let args = vec![("x".to_string(), Ty::Int)];
    let _ = define_function(&mut resolver, "test_func", args, Ty::String);

    // Snapshot the resolver state
    assert_snapshot!(format!("STATE: {:#?}", resolver));
}

#[test]
fn function_redefinition_error() {
    let mut resolver = Resolver::new();

    let args = vec![("x".to_string(), Ty::Int)];
    let _ = define_function(&mut resolver, "test_func", args, Ty::String);

    // Attempt to redefine the function with a different signature
    let args = vec![("y".to_string(), Ty::String)];
    let result = define_function(&mut resolver, "test_func", args, Ty::Int);

    // Snapshot the result and resolver state
    assert_error!(result, SemanticError::RedefinedVariable(_));
}

#[test]
fn valid_function_scoping() {
    let mut resolver = Resolver::new();

    let func_ty = Ty::Fn(Box::new(crate::ast::ast_types::Function {
        name: "global_func".to_string(),
        args: vec![],
        return_ty: Ty::Unit,
        body: None,
        body_expr: None,
    }));

    // Define function in global scope
    resolver
        .define_function("global_func".to_string(), func_ty.clone())
        .unwrap();

    // Push new scope
    resolver.push_scope();

    // Define local function
    let local_func_ty = Ty::Fn(Box::new(crate::ast::ast_types::Function {
        name: "local_func".to_string(),
        args: vec![],
        return_ty: Ty::Int,
        body: None,
        body_expr: None,
    }));

    resolver
        .define_function("local_func".to_string(), local_func_ty.clone())
        .unwrap();

    // Pop scope
    resolver.pop_scope().unwrap();

    // Snapshot the resolver state
    assert_snapshot!(format!("STATE:\n---\n: {:#?}", resolver));
}
