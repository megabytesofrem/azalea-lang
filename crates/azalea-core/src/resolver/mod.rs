use crate::resolver::semantic_error::SemanticError;

pub mod resolver;
pub mod semantic_error;

pub mod resolver_tests;

pub type Return<'a, T> = Result<T, SemanticError>;
