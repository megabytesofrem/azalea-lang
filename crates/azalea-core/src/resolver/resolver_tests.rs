use crate::ast::ast_types::Ty;
use crate::resolver::resolver::Resolver;
use crate::resolver::semantic_error::SemanticError;
use insta::assert_snapshot;

#[test]
fn basic_variable_definition() {
    let mut resolver = Resolver::new();

    // Define a variable in global scope
    resolver.define_variable("x".to_string(), Ty::Int).unwrap();

    // Snapshot the resolver state
    assert_snapshot!(format!("{:#?}", resolver));
}

#[test]
fn errors_on_redefinition() {
    let mut resolver = Resolver::new();

    resolver.define_variable("x".to_string(), Ty::Int).unwrap();

    // Attempt to redefine the variable
    let result = resolver.define_variable("x".to_string(), Ty::String);

    // Snapshot the result and resolver state
    assert_snapshot!(format!("Result: {:#?}\nResolver: {:#?}", result, resolver));
}

#[test]
fn valid_nested_scopes() {
    let mut resolver = Resolver::new();

    // Define variable in global scope
    resolver.define_variable("x".to_string(), Ty::Int).unwrap();

    // Push new scope
    resolver.push_scope();

    // Define new variable in inner scope
    resolver
        .define_variable("y".to_string(), Ty::String)
        .unwrap();

    // Pop scope
    resolver.pop_scope().unwrap();

    // Snapshot the resolver state
    assert_snapshot!(format!("{:#?}", resolver));
}

#[test]
fn variable_shadowing() {
    let mut resolver = Resolver::new();

    // Define variable in global scope
    resolver.define_variable("x".to_string(), Ty::Int).unwrap();

    // Push new scope and shadow the variable
    resolver.push_scope();
    resolver
        .define_variable("x".to_string(), Ty::String)
        .unwrap();

    // Pop scope
    resolver.pop_scope().unwrap();

    // Snapshot the resolver state
    assert_snapshot!(format!("{:#?}", resolver));
}

#[test]
fn does_allow_redefinition() {
    let mut resolver = Resolver::new();

    resolver.define_variable("x".to_string(), Ty::Int).unwrap();

    // Redefine the variable
    resolver.define_or_redefine_variable("x".to_string(), Ty::String);

    // Snapshot the resolver state
    assert_snapshot!(format!("{:#?}", resolver));
}

#[test]
fn errors_on_empty_scope() {
    let mut resolver = Resolver::new();

    // Attempt to pop the global scope
    let result = resolver.pop_scope();

    // Snapshot the result and resolver state
    assert_snapshot!(format!("Result: {:#?}\nResolver: {:#?}", result, resolver));
}

#[test]
fn undefined_variable() {
    let resolver = Resolver::new();

    // Snapshot the lookup result
    assert_snapshot!(format!(
        "Lookup: {:#?}",
        resolver.lookup_variable("undefined")
    ));
}

#[test]
fn function_definition_and_lookup() {
    let mut resolver = Resolver::new();

    // Define a function in global scope
    let func_ty = Ty::Fn(Box::new(crate::ast::ast_types::Function {
        name: "test_func".to_string(),
        args: vec![("x".to_string(), Ty::Int)],
        return_ty: Ty::String,
        body: None,
        body_expr: None,
    }));

    resolver
        .define_function("test_func".to_string(), func_ty.clone())
        .unwrap();

    // Snapshot the resolver state
    assert_snapshot!(format!("{:#?}", resolver));
}

#[test]
fn function_redefinition_error() {
    let mut resolver = Resolver::new();

    let func_ty1 = Ty::Fn(Box::new(crate::ast::ast_types::Function {
        name: "test_func".to_string(),
        args: vec![("x".to_string(), Ty::Int)],
        return_ty: Ty::String,
        body: None,
        body_expr: None,
    }));

    let func_ty2 = Ty::Fn(Box::new(crate::ast::ast_types::Function {
        name: "test_func".to_string(),
        args: vec![("y".to_string(), Ty::String)],
        return_ty: Ty::Int,
        body: None,
        body_expr: None,
    }));

    resolver
        .define_function("test_func".to_string(), func_ty1)
        .unwrap();

    // Attempt to redefine the function
    let result = resolver.define_function("test_func".to_string(), func_ty2);

    // Snapshot the result and resolver state
    assert_snapshot!(format!("Result: {:#?}\nResolver: {:#?}", result, resolver));
}

#[test]
fn valid_function_scoping() {
    let mut resolver = Resolver::new();

    let func_ty = Ty::Fn(Box::new(crate::ast::ast_types::Function {
        name: "global_func".to_string(),
        args: vec![],
        return_ty: Ty::Unit,
        body: None,
        body_expr: None,
    }));

    // Define function in global scope
    resolver
        .define_function("global_func".to_string(), func_ty.clone())
        .unwrap();

    // Push new scope
    resolver.push_scope();

    // Define local function
    let local_func_ty = Ty::Fn(Box::new(crate::ast::ast_types::Function {
        name: "local_func".to_string(),
        args: vec![],
        return_ty: Ty::Int,
        body: None,
        body_expr: None,
    }));

    resolver
        .define_function("local_func".to_string(), local_func_ty.clone())
        .unwrap();

    // Pop scope
    resolver.pop_scope().unwrap();

    // Snapshot the resolver state
    assert_snapshot!(format!("{:#?}", resolver));
}
