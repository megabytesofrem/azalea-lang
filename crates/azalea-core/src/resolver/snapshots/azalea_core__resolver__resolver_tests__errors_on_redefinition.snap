---
source: crates/azalea-core/src/resolver/resolver_tests.rs
expression: "format!(\"Result: {:#?}\\nResolver: {:#?}\", result, resolver)"
---
Result: Err(
    RedefinedVariable(
        "x",
    ),
)
Resolver: Resolver {
    scopes: [
        Scope {
            variables: {
                "x": Int,
            },
            functions: {},
        },
    ],
}
